### ISSUE TITLE
DD memory footprint after purging all inputs

### ISSUE THREAD

**Issue by ryzhyk** [TimelyDataflow/differential-dataflow#188]
I have a DD computation that feeds many records to the program, propagates them through the DD graph to compute output updates, then clears all input relations and again propagates changes through the graph (at which point all collections are empty).  It does this three times, hence the three spikes on the following memory profile computed by `valgrind --tool=massif`:

```
    MB
924.4^                                                               #        
     |                                                               #:       
     |                                      @@:                      #@       
     |                                     @@@:                   ::@#@@      
     |               ::                  @@@@@@:             :  :::@@#@@@:    
     |              :@:@            :    @@@@@@::          :@: @@@@@@#@@@:    
     |             ::@:@: :        :: @@:@@@@@@@@@::      ::@@@@@@@@@#@@@@@@: 
     |            :::@:@@::       :@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |            :::@:@@:::@@    :@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |          @::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |         :@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |        ::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |       :::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |      :@::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |     @@@::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |     @@@::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |     @@@::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |    :@@@::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |    @@@@::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
     |@@:@@@@@::@::::@:@@:::@@:@:@:@:@@@:@@@@@@@@@:::@::@@@:@@@@@@@@@#@@@@@@@@
   0 +----------------------------------------------------------------------->Gi
     0                                                                   124.9
```

What worries me is that the valleys in the graph, when all relations are empty, are nowhere near close to zero.  Just to clarify: all input handles advanced to the next timestamp at this point, so DD should know not to expect updates to past timestamps.

Is this expected behavior, caused by how DD compacts history?  If so, is there a way to force DD to cleanup the old state?

**Comment by ryzhyk** [comment]
Here is how the size of an arranged output collection changes as we were populate and purge inputs: 

```
Input collections populated with data             567594
Input relations purged (output is empty)          1135188
Input collections re-populated from scratch       1702782 // we now store 3x the  number of entries in the collection
Input relations purged (ouput is empty)           1135188
Input collections re-populated from scratch       1702782 // arrangements stop growing after three iterations
Input relations purged:                           1135188
```

**Comment by frankmcsherry** [comment]
This is "expected" behavior, though I can understand why you might neither expect nor want it. Let me talk through what is going on, and present some of the trade-offs we have to deal with. I'm up for thinking through appropriate fixes or amendments, but they pull and tug against other constraints.

Trace maintenance is performed in response to additional updates that arrive, with the intent of rate limiting the amount of maintenance work we perform. The intuition is that we should not do substantially more work than changes that have arrived, to maintain some latency properties proportionate to the amount of change flowing around. In particular, merging of LSM layers happens in proportion to new updates received.

One consequence of this is that when updates stop flowing, maintenance work stops happening. In some eyes, this is ideal. For example, other people could plausibly be stressed out when the stop supplying updates and yet the system is still grinding away, profitably compacting their gigabytes of data in the background.

We could pretty easily make `arrange` continue to re-schedule itself whenever it has pending merge work, and perform some minimal amount of merge work each time it is scheduled. This would introduce some amount of CPU even when changes are not underway, but it would eventually quiesce is some state with likely reduced memory footprint, even for people who don't expect their data to cancel (each merge in progress will release memory when complete).

This doesn't guarantee that a collection that is empty should have an empty memory footprint, because there can also be merges that have not been initiated because LSM layers have appropriately geometrically decreasing sizes. To move in this direction, we would need to initiate merges that are not mathematically justified (not two layers with sizes within the factor of two / scale factor). We could do that by having `arrange` introduce bogus "empty" batches that are easy to merge, but adds "fuel" to the merge process. It is less clear on what basis we should do this. The `arrange` operator could continually reschedule itself when it has any more than a single LSM layer, adding some constant amount of merge fuel with each invocation, which will eventually provoke the initiation of non-trivial merges, and which should eventually result in a single LSM layer, at the cost of a fair bit of unexpected CPU.

This sounds like good fodder for constraints / desiderata of a state management rework. I have a branch with this in progress, for other reasons (one you won't like: some other people prefer to retain the memory to avoid thrashing the system allocator). But it might be a good place to talk out the design and see if perhaps we can expose some more knobs from `arrange` so that you can invoke a more memory-lean version if you prefer, and others could invoke a more CPU version if they prefer.

**Comment by ryzhyk** [comment]
Thanks for the detailed explanation! This is roughly how I imagined this to work and I can understand the motivation behind this design.  Unfortunately, our use cases seem to have very different performance requirements.  Differential's incremental performance is amazing; in fact if incremental updates were 10x slower or occasionally took even longer due to compaction, noone would complain.  The 2x or 3x memory overhead, on the other hand, is an absolute non-starter (in fact even the current 1x is problematic and we are trying hard to reduce it). 

Sounds like there is no one-size-fit-all solution here, so exposing control knobs to users to trigger compaction seems like the best long-term solution.

Short-term, I wonder if there is a way to generate input to DD to trigger compaction?

**Comment by frankmcsherry** [comment]
> Short-term, I wonder if there is a way to generate input to DD to trigger compaction?

I think differential is too smart, and will vanish any work that does not exist, and work that exists probably (maybe?) won't flush out your state. You could try alternately introducing `(x, now, 1)` and `(x, now+1, -1)` and see if it helps (it could, but it could also gum up the works more until it does).

My guess as to the best medium term strategy is for me to allow the `arrange` operator to be constructed with a parameter that will have it continually re-activate as long as the trace it manages has outstanding merges to perform. This almost seems like a no brainer, as this is work that has to be done (unless the program terminates first) and the only overhead is some additional but manageable CPU burden while running. As an additional bonus, once these merges complete reads become faster as there are fewer LSM layers to navigate.

I'm off for a few days, but I'll try and spec this out and see what it looks like. It doesn't sound too hard, and in the worst case we can prototype it with an environment variable that you can set.

**Comment by frankmcsherry** [comment]
I've set up a branch to start looking into this. Unfortunately, while I've got the machinery to more pro-actively merge in place, that doesn't seem to be the root of the problem. At least, I've written a program (`examples/compact.rs`) that loads and then unloads a fair bit of data.  Looking at the LSM layers, they are almost always not in a "merging" state, including none of them at the end. There are nonetheless several complete merges, apparently representing an empty collection.

Due to the geometric sizing of batches, and that the negations are as many in number as the additions, the largest batch must have both additions and negations in it, which means that it is due for a compaction (the frontier has advanced) but no other batch has challenged it. This is a bit of an weird situation, because it is hard for us to predict whether the batch will shrink if we start a compaction (we only perform them as parts of merges, rather than proactively when frontiers move).

It looks like, at least for my example, we will need to get to introducing phantom updates, as a way to continually provoke merging and such, even without further update inputs or frontier advances.
